File: examples/go/case04_test.go
package examplesgo

import "testing"

func TestCase04DeliveryAndProtectedEffect(t *testing.T) {
	weak, weakAck := NewRewardStore(), &DeliveryAck{Accepted: map[string]bool{}}
	WeakEffectThenAck(weak, weakAck, "E", "User42", 100, true)
	if weakAck.Accepted["E"] {
		t.Fatal("the pre-ACK crash was acknowledged")
	}
	WeakEffectThenAck(weak, weakAck, "E", "User42", 100, false)
	if weak.State.Balance["User42"] != 200 {
		t.Fatal("the weak D1+D2 trace did not repeat the reward")
	}

	ackFirst, firstAck := NewRewardStore(), &DeliveryAck{Accepted: map[string]bool{}}
	AckBeforeEffect(ackFirst, firstAck, "E", "User42", 100, true)
	if !firstAck.Accepted["E"] || ackFirst.State.Balance["User42"] != 0 {
		t.Fatal("ACK-first did not expose the lost-effect gap")
	}

	strong, strongAck := NewRewardStore(), &DeliveryAck{Accepted: map[string]bool{}}
	if !strong.ApplyEventOnce("E", "User42", 100) {
		t.Fatal("first-time E was suppressed")
	}
	if strong.ApplyEventOnce("E", "User42", 100) || strong.State.Balance["User42"] != 100 {
		t.Fatal("D2(E) repeated the protected reward")
	}
	HandleRewardDelivery(strong, strongAck, "E", "User42", 100)
	if !strongAck.Accepted["E"] || strong.State.Balance["User42"] != 100 {
		t.Fatal("redelivery could not finish safely")
	}

	split, marker := NewRewardStore(), map[string]bool{}
	SplitRewardAndMarker(split, marker, "E", "User42", 100, true)
	SplitRewardAndMarker(split, marker, "E", "User42", 100, false)
	if split.State.Balance["User42"] != 200 {
		t.Fatal("split marker did not expose the repeated effect")
	}
}

File: examples/java/Case04Test.java
import java.util.HashSet;
import java.util.Set;

final class Case04Test {
    private static void require(boolean condition, String message) {
        if (!condition) throw new AssertionError(message);
    }

    public static void main(String[] args) {
        var weak = new Case04.RewardStore();
        var weakAck = new Case04.DeliveryAck();
        Case04.weakEffectThenAck(weak, weakAck, "E", "User42", 100, true);
        require(!weakAck.accepted.contains("E"), "pre-ACK crash was acknowledged");
        Case04.weakEffectThenAck(weak, weakAck, "E", "User42", 100, false);
        require(weak.state.balance.get("User42") == 200, "weak D1+D2 did not repeat");

        var ackFirst = new Case04.RewardStore();
        var firstAck = new Case04.DeliveryAck();
        Case04.ackBeforeEffect(ackFirst, firstAck, "E", "User42", 100, true);
        require(firstAck.accepted.contains("E") && !ackFirst.state.balance.containsKey("User42"),
                "ACK-first did not expose lost effect");

        var strong = new Case04.RewardStore();
        require(Case04.applyEventOnce(strong, "E", "User42", 100), "new E suppressed");
        require(!Case04.applyEventOnce(strong, "E", "User42", 100), "D2(E) applied");
        var strongAck = new Case04.DeliveryAck();
        Case04.handleRewardDelivery(strong, strongAck, "E", "User42", 100);
        require(strong.state.balance.get("User42") == 100 && strongAck.accepted.contains("E"),
                "redelivery did not finish safely");

        var split = new Case04.RewardStore();
        Set<String> marker = new HashSet<>();
        Case04.splitRewardAndMarker(split, marker, "E", "User42", 100, true);
        Case04.splitRewardAndMarker(split, marker, "E", "User42", 100, false);
        require(split.state.balance.get("User42") == 200, "split marker did not repeat effect");
    }
}
